src/mpid/ch4/shm/posix/eager/fbox/fbox_send.h

```diff
@@ -108,7 +108,7 @@ MPIDI_POSIX_eager_send(int grank,
     }
 #endif /* POSIX_FBOX_DEBUG */
 
-    OPA_compiler_barrier();
```

Comment:
This translation is correct, but I don't think this is necessary. It just avoids instruction reordering in assembly level and does not have any effect on data visibility from other threads.

If they are not in critical paths, read-write barrier can be put instead. Otherwise, we need to understand the intention of these compiler barriers.

Reply:
If the statement below (`fbox_out->data_ready = 1`) gets re-ordered and the other process sees `data_ready` before early statements that set the content, the other process will be reading the wrong data. Here `data_ready` is typed as `volatile uint64_t`. Maybe the proper way is to type it as `MPL_atomic_int_t` and use `acquire/release` rather than the ` compiler_barrier`?

Reply:
This is in a very performance critical part. Is there performance benefit with `volatile` plus `compiler_barrier` vs `acquire/release` or whether the former guarantee correctness?

Reply:
Your idea in the first comment sounds great; generally, it's a good idea to use atomic variables for any synchronization variables (in this case, `data_ready`) and avoid read/write barriers. It should be optimal.

Reply:
Sounds good. I'll make the change.